### ISSUE TITLE
Question: In P4_14, why are `header_ref.valid` and `header_ref` both keys?

### ISSUE THREAD

**Issue by cole-barefoot** [p4lang/p4-spec#413]
There seem to be two ways to check the validity of a header in a table:

```p4
header some_type_t h;

table t1 {
  reads { h.valid : exact; }
  ...
}

table t2 {
  reads { h : valid; }
  ...
}
```

The former is undocumented and only appears in the BNF rule for table reads (Section 11), from what I can tell.  Assuming the intention is to provide access to the validity bit associated with the header, then `h.valid : exact` is equivalent to `h : valid`, but also supports more functionality, eg. `h.valid : ternary`.

Why do both exist?  Can we remove the `valid` match type in favor of the `valid` pseudo-field?  If not, perhaps the documentation in Section 11 (and/or 2.2.1?) can be updated to explain the two different ways to check validity.

> EDIT: Change examples to use P4_14 syntax (`reads`) rather than P4_16 syntax (`keys`).

**Comment by mbudiu-vmw** [comment]
Historically, for a long time only the second one existed.
This is what P4-16 did: remove the second one and only use the first one.
The second one must exist for backwards compatibility.

**Comment by jnfoster** [comment]
@cole-barefoot your question seems to mix P4_14 and P4_16. 

In the context of P4_14, As @mbudiu-vmw indicated, the `.valid` pseudo-header was only added in the past year (version 1.04 IIRC?). It seemed dangerous to remove the `valid` match kind since that might break existing programs.

The standard library for P4_16 (as defined in `core.p4`) doesn't have a `valid` `match_kind`.

**Comment by cole-barefoot** [comment]
No, this is for the P4_14 spec.  Sorry, I should have specified that in the issue title, which I've now edited.

Reading through the [P4_14 spec](https://p4lang.github.io/p4-spec/p4-14/v1.0.4/tex/p4.pdf), Section 11 (Table Declarations) shows a grammar fragment that accepts both `h.valid : exact` and `h : valid`.

The change log in that document shows that the `.valid` pseudo-field was added in v1.0.4 (see Section 15.3.1).

I don't see any other mention of the `.valid` pseudo-field in the P4_14 spec.
